mq.go

```diff
@@ -137,48 +206,145 @@ func (mq *MQ) connect() error {
 		return err
 	}
 
+	if err := channel.Qos(prefetchCount, prefetchSize, false); err != nil {
+		channel.Close()
+		connection.Close()
+
+		return err
+	}
+
 	mq.connection = connection
 	mq.channel = channel
 
+	go mq.handleCloseEvent()
+
 	return nil
 }
 
+// Register close handler.
+// To more details visit https://godoc.org/github.com/streadway/amqp#Connection.NotifyClose.
+func (mq *MQ) handleCloseEvent() {
+	err := <-mq.connection.NotifyClose(make(chan wabbit.Error))
+	if err != nil {
+		mq.internalErrorChannel <- err
+	}
+}
+
 func (mq *MQ) errorHandler() {
 	for err := range mq.internalErrorChannel {
 		select {
-		case mq.errorChannel <- err:
-			mq.processError(err)
+		case mq.errorChannel <- err: // Proxies errors to the user.
 		default: // Drop errors if channel buffer is full.
+			// TODO probably it makes sense to make it optional or even remove.
 		}
+
+		mq.processError(err)
 	}
 }
 
 func (mq *MQ) processError(err interface{}) {
-	if rmqErr, ok := err.(*amqpDriver.Error); ok {
+	switch err.(type) {
+	case *net.OpError:
+		go mq.reconnect()
+	case *amqpDriver.Error:
+		rmqErr, _ := err.(*amqpDriver.Error)
 		if rmqErr.Server == false { // For example channel was closed.
-			time.Sleep(mq.reconnectTimeout)
-			mq.reconnect()
+			go mq.reconnect()
 		}
+	default:
+		// There is no special behaviour for other errors.
 	}
 }
 
+// Reconnect stops current producers and consumers,
+// recreates connection to the rabbit and than runs producers and consumers.
 func (mq *MQ) reconnect() {
-	mq.connect()
-	mq.producers.SetNewChannel(mq.channel)
+	notBusy := atomic.CompareAndSwapInt32(&mq.reconnectStatus, statusReadyForReconnect, statusReconnecting)
+	if !notBusy {
+		// There is no need to start new reconnect if previous is not finished yet.
```

Comment:
There is no need to start a new reconnect if the previous one is not finished yet.